src/lib/distances.jl

```diff
@@ -43,28 +44,28 @@ end
   function(Δ)
     d1 = Diagonal(vec(sum(Δ; dims=1)))
     d2 = Diagonal(vec(sum(Δ; dims=2)))
-    return (nothing, x * (2 .* (d1 .+ d2 .- Δ .- transpose(Δ))) |> f)
+    return NO_FIELDS, NO_FIELDS, x * (2 .* (d1 .+ d2 .- Δ .- transpose(Δ))) |> f
   end
 
-@adjoint function (::Euclidean)(x::AbstractVector, y::AbstractVector)
+function rrule(::Euclidean, x::AbstractVector, y::AbstractVector)
   D = x .- y
   δ = sqrt(sum(abs2, D))
   function euclidean(Δ::Real)
     x̄ = ifelse(iszero(δ), D, (Δ / δ) .* D)
-    return x̄, -x̄
+    return NO_FIELDS, x̄, -x̄
   end
   return δ, euclidean
 end
 
-@adjoint function colwise(s::Euclidean, x::AbstractMatrix, y::AbstractMatrix)
+function rrule(::typeof(colwise), s::Euclidean, x::AbstractMatrix, y::AbstractMatrix)
   d = colwise(s, x, y)
   return d, function (Δ::AbstractVector)
     x̄ = (Δ ./ max.(d, eps(eltype(d))))' .* (x .- y)
-    return nothing, x̄, -x̄
+    return NO_FIELDS, NO_FIELDS, x̄, -x̄
   end
 end
 
-@adjoint function pairwise(::Euclidean, X::AbstractMatrix, Y::AbstractMatrix; dims=2)
+function rrule(::typeof(pairwise), ::Euclidean, X::AbstractMatrix, Y::AbstractMatrix; dims=2)
```

Comment:
This definition calls `Zygote.pullback` to compute the pullback of a more stable modification of the forward pass. This seems unfortunate for everyone that would like to use the ChainRules definitions with another AD backend. But on the other hand, the definitions here are type piracy anyway and problematic in such cases where Zygote is loaded (possibly by some other dependency).

Ref: https://github.com/JuliaDiff/ChainRules.jl/issues/338, https://github.com/JuliaDiff/ChainRulesCore.jl/issues/68

Reply:
@theogf up to revive this using the pattern mentioned in https://github.com/JuliaDiff/ChainRules.jl/issues/338#issuecomment-884772928?

Reply:
Done! Hopefully I understood correctly how it worked